### PR TITLE
avm2: Improve Stage3D support for textures, register types, and opcodes

### PR DESCRIPTION
**Pull request by Aaron1011**
This is a very large diff, but most of it comes from test files and output.

This PR ads partial support for the following Stage3D shader features:
* Normal (square), rectangle, and cube textures
* Varying and temporary registers
* Lots of opcodes

The combination of these allows us to get a raytracing program fully working in Ruffle. I've included it as image test. Currently, this test is very slow (about 90 seconds on my machine), as the code I'm using (https://github.com/saharan/OGSL) includes its own shader language and compiler. THe raytracing demo first compiles its own shader language to AGAL, and then starts rendering the scene.

Limitations:
* Many opcodes are still unimplemented
* Most non-default texture options (e.g. mipmaps) are not implemented

Here's the raytracing result:

expected

### CODE REVIEW

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 1]
```diff
@@ -298,3 +369,117 @@ pub fn clear<'gc>(
...
+        // This is a native method, so all of the arguments have been checked and coerced for us
+        let width = args[0].as_integer(activation.context.gc_context)? as u32;
+        let height = args[1].as_integer(activation.context.gc_context)? as u32;
+        let format = args[2].coerce_to_string(activation)?;
+        let optimize_for_render_to_texture = args[3].coerce_to_boolean();
+        let streaming_levels = args[4].as_integer(activation.context.gc_context)? as u32;
+        let format = if &*format == b"bgra" {
+            Context3DTextureFormat::Bgra
+        } else {
+            panic!("Unsupported texture format: {:?}", format);
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display3D/context_3d.rs:388]: Is this just for visibility?

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 2]
```diff
@@ -298,3 +369,117 @@ pub fn clear<'gc>(
...
+    if let Some(context) = this.and_then(|this| this.as_context_3d()) {
+        // This is a native method, so all of the arguments have been checked and coerced for us
+        let width = args[0].as_integer(activation.context.gc_context)? as u32;
+        let height = args[1].as_integer(activation.context.gc_context)? as u32;
+        let format = args[2].coerce_to_string(activation)?;
+        let optimize_for_render_to_texture = args[3].coerce_to_boolean();
+        let format = if &*format == b"bgra" {
+            Context3DTextureFormat::Bgra
+        } else {
+            panic!("Unsupported texture format: {:?}", format);
```
Comment (Dinnerbone) [inline core/src/avm2/globals/flash/display3D/context_3d.rs:420]: Maybe give them different messages if so :D

File: core/src/avm2/globals/flash/display3D/context_3d.rs (hunk 1 above)
Comment (Aaron1011) [reply core/src/avm2/globals/flash/display3D/context_3d.rs:388]: Yes - Stage3D support isn't complete enough yet for this to be a normal error.

File: tests/tests/swfs/avm2/stage3d_raytrace/test.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+[image_comparison]
+tolerance = 0
+# FIXME - see if we can decrease this. Ideally, it wouldn't be nearly so high
+max_outliers = 5100
+
+[player_options]
+# This test runs very slowly on unoptimized `test` builds,
+# so give it plenty of time.
+max_execution_duration = { secs = 1000, nanos = 0 }
+with_renderer = { optional = false, sample_count = 1 }
```
Comment (Dinnerbone) [inline tests/tests/swfs/avm2/stage3d_raytrace/test.toml:16]: I think renderer would be required for this, no?

**Comment by Dinnerbone** [comment]
Nit: You dropped the prefixes on most commits

**Comment by Aaron1011** [comment]
@Dinnerbone: I've planning to squash them all into the first commit.

File: tests/tests/swfs/avm2/stage3d_raytrace/test.toml (hunk 3 above)
Comment (Aaron1011) [reply tests/tests/swfs/avm2/stage3d_raytrace/test.toml:16]: That's right - this is `optional = false`, so it's required.

Comment (Dinnerbone) [reply tests/tests/swfs/avm2/stage3d_raytrace/test.toml:16]: Oh gosh sorry, misread it!

**Review by Dinnerbone — APPROVED** [review]

**Comment by Adamillo** [comment]
I do wonder if this gets Super Fancy Pants Adventure (or World 4 with 3 parts) to boot at all?

**Comment by Aaron1011** [comment]
They both need more XML support, unfortunately

File: core/src/avm2/globals.rs [hunk 4]
```diff
@@ -209,14 +212,17 @@ impl<'gc> SystemClasses<'gc> {
...
             bitmapfilter: object,
-            blurfilter: object,
             colormatrixfilter: object,
             convolutionfilter: object,
             displacementmapfilter: object,
             dropshadowfilter: object,
             glowfilter: object,
             gradientbevelfilter: object,
             gradientglowfilter: object,
+            blurfilter: object,
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals.rs:222]: This looks like it was moved accidentally, it probably shouldn't be moved.

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 5]
```diff
@@ -298,3 +369,123 @@ pub fn clear<'gc>(
...
+    if let Some(context) = this.and_then(|this| this.as_context_3d()) {
+        // This is a native method, so all of the arguments have been checked and coerced for us
+        let size = args[0].as_integer(activation.context.gc_context)? as u32;
+        let format = args[1].coerce_to_string(activation)?;
+        let optimize_for_render_to_texture = args[2].coerce_to_boolean();
+        let streaming_levels = args[3].as_integer(activation.context.gc_context)? as u32;
+        let format = if &*format == b"bgra" {
+            Context3DTextureFormat::Bgra
+        } else {
+            panic!(
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display3D/context_3d.rs:455]: Should this be a panic?

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 6]
```diff
@@ -298,3 +369,123 @@ pub fn clear<'gc>(
...
+    if let Some(context) = this.and_then(|this| this.as_context_3d()) {
+        // This is a native method, so all of the arguments have been checked and coerced for us
+        let width = args[0].as_integer(activation.context.gc_context)? as u32;
+        let height = args[1].as_integer(activation.context.gc_context)? as u32;
+        let format = args[2].coerce_to_string(activation)?;
+        let optimize_for_render_to_texture = args[3].coerce_to_boolean();
+        let format = if &*format == b"bgra" {
+            Context3DTextureFormat::Bgra
+        } else {
+            panic!(
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display3D/context_3d.rs:420]: This shouldn't be a panic

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 7]
```diff
@@ -280,6 +288,69 @@ pub fn set_program_constants_from_matrix<'gc>(
...
+            .get(0)
+            .unwrap_or(&Value::Undefined)
+            .coerce_to_string(activation)?;
+
+        let program_type = if &*program_type == b"vertex" {
+            ProgramType::Vertex
+        } else if &*program_type == b"fragment" {
+            ProgramType::Fragment
+        } else {
+            panic!("Unknown program type {:?}", program_type);
```
Comment (Lord-McSweeney) [inline core/src/avm2/globals/flash/display3D/context_3d.rs:307]: This shouldn't be a panic.

**Comment by Lord-McSweeney** [comment]
I see a lot of `panic!`s on unimplemented features, those should probably be removed.

File: core/src/avm2/globals/flash/display3D/context_3d.rs [hunk 8]
```diff
@@ -298,3 +369,117 @@ pub fn clear<'gc>(
...
 }
+
+pub fn create_texture<'gc>(
+    activation: &mut Activation<'_, 'gc>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error<'gc>> {
+    if let Some(context) = this.and_then(|this| this.as_context_3d()) {
+        // This is a native method, so all of the arguments have been checked and coerced for us
+        let width = args[0].as_integer(activation.context.gc_context)? as u32;
```
Comment (adrian17) [inline core/src/avm2/globals/flash/display3D/context_3d.rs:380]: future nit: `as_integer` is still relatively heavyweight (calls `value_of` + error text formatting), so might wanna smth even simpler if the arguments were checked

File: core/src/avm2/globals/json.rs [hunk 9]
```diff
@@ -118,8 +118,9 @@ impl<'gc> AvmSerializer<'gc> {
         } else {
             let obj = value.as_object().unwrap();
             let to_json = obj
-                .get_public_property("toJSON", activation)?
-                .as_object()
+                .get_public_property("toJSON", activation)
+                .ok()
```
Comment (adrian17) [inline core/src/avm2/globals/json.rs:122]: can this be split?

**Review by adrian17 — APPROVED** [review]

**Pull request merged**